src/java.base/share/classes/java/security/Provider.java

```diff
@@ -1557,12 +1557,22 @@ private static class EngineDescription {
         final String name;
         final boolean supportsParameter;
         final String constructorParameterClassName;
+        private volatile Class<?> constructorParameterClass;
 
         EngineDescription(String name, boolean sp, String paramName) {
             this.name = name;
             this.supportsParameter = sp;
             this.constructorParameterClassName = paramName;
         }
+
+        Class<?> getConstructorParameterClass() throws ClassNotFoundException {
```

Comment:
I think we want to do what `Service.getImplClass` is doing, for extra safety: retain cached class on a `WeakReference`. This would avoid accidental class leakage in this code.

Reply:
Thanks. I reverted it exactly from JDK-8280970 because a) that particular code was there for a long time and presumably was reasonably well used/tested without big issues and b) I want to backport the optimization to JDK17 which still has this exact implementation.

Aside: The classes saved here are limited to the 31 explicitly added in Provider.<clinit>. I'm not sure if that helps limit the leak potential significantly?

Reply:
> Aside: The classes saved here are limited to the 31 explicitly added in Provider.<clinit>. I'm not sure if that helps limit the leak potential significantly?

True. Might not even be an issue in practice.

I think the argument for keeping the code in pre-JDK-8280970 form is a good move for future backports.